File: apps/mobile/lib/core/pitch.dart
// Mic pitch capture (flutter_audio_capture → Float32 PCM → pitch_detector_dart
// YIN) and shape-scoring for the "Your turn" maqām practice. We compare the
// SHAPE of the learner's melody to the reference (normalized + progress-aligned
// + correlation), so absolute pitch and octave don't matter — only the
// up-and-down journey.

import 'dart:async';
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter_audio_capture/flutter_audio_capture.dart';
import 'package:permission_handler/permission_handler.dart';
import 'package:pitch_detector_dart/pitch_detector.dart';

const int _bufSize = 2048;
const int _sampleRate = 44100;

class PitchSession {
  final _capture = FlutterAudioCapture();
  final _detector = PitchDetector(audioSampleRate: _sampleRate * 1.0, bufferSize: _bufSize);
  final List<double> _buf = [];
  bool _running = false;

  /// Starts the mic; calls [onPitch] with each detected Hz. Returns null on
  /// success, or a user-facing message explaining why it couldn't start —
  /// distinguishing a denied permission from a mic that wouldn't open.
  Future<String?> start(void Function(double hz) onPitch) async {
    final status = await Permission.microphone.request();
    if (!status.isGranted) {
      return 'Microphone access is needed for “Your turn”. Please allow it and try again.';
    }
    try {
      await _capture.start(
        (dynamic obj) async {
          final samples = obj as Float32List;
          _buf.addAll(samples);
          while (_buf.length >= _bufSize) {
            final window = _buf.sublist(0, _bufSize);
            _buf.removeRange(0, _bufSize);
            try {
              final res = await _detector.getPitchFromFloatBuffer(window);
              if (res.pitched && res.pitch >= 70 && res.pitch <= 700) onPitch(res.pitch);
            } catch (_) {/* skip bad window */}
          }
        },
        (Object e) {/* stream error — ignore, session just yields no pitch */},
        sampleRate: _sampleRate,
        bufferSize: _bufSize,
      );
    } catch (_) {
      return 'Couldn’t open the microphone. Close other apps that might be using it and try again.';
    }
    _running = true;
    return null;
  }

  Future<void> stop() async {
    if (!_running) return;
    _running = false;
    try {
      await _capture.stop();
    } catch (_) {}
    _buf.clear();
  }

  void dispose() {
    if (_running) _capture.stop();
  }
}

double _log2(double x) => log(x) / ln2;

/// Map a pitch in Hz to a 0..1 ribbon height over a plausible vocal range.
double pitchToLevel(double hz) {
  final lo = _log2(90), hi = _log2(520);
  return ((_log2(hz) - lo) / (hi - lo)).clamp(0.0, 1.0);
}

List<double> _resample(List<double> src, int n) {
  if (src.isEmpty) return List.filled(n, 0);
  if (src.length == 1) return List.filled(n, src.first);
  final out = <double>[];
  for (var i = 0; i < n; i++) {
    final t = i / (n - 1) * (src.length - 1);
    final lo = t.floor(), hi = min(src.length - 1, lo + 1);
    out.add(src[lo] + (src[hi] - src[lo]) * (t - lo));
  }
  return out;
}

List<double> _minMax(List<double> a) {
  final lo = a.reduce(min), hi = a.reduce(max);
  if (hi - lo < 1e-6) return a.map((_) => 0.5).toList();
  return a.map((v) => (v - lo) / (hi - lo)).toList();
}

double _pearson(List<double> a, List<double> b) {
  final n = min(a.length, b.length);
  var ma = 0.0, mb = 0.0;
  for (var i = 0; i < n; i++) {
    ma += a[i];
    mb += b[i];
  }
  ma /= n;
  mb /= n;
  var num = 0.0, da = 0.0, db = 0.0;
  for (var i = 0; i < n; i++) {
    final x = a[i] - ma, y = b[i] - mb;
    num += x * y;
    da += x * x;
    db += y * y;
  }
  if (da < 1e-9 || db < 1e-9) return 0;
  return num / sqrt(da * db);
}

/// How closely the learner's melodic SHAPE followed the reference (0..100).
int shapeMatch(List<double> userHz, List<double> refLevels) {
  if (userHz.length < 6) return 0;
  const n = 64;
  final user = _minMax(_resample(userHz.map(_log2).toList(), n));
  final ref = _resample(refLevels, n);
  return (max(0.0, _pearson(user, ref)) * 100).round();
}

File: apps/web/lib/audio/pitch.ts
// Mic pitch capture (Web Audio + autocorrelation) and shape-scoring helpers for
// the "Your turn" maqām practice. We compare the SHAPE of the learner's melody
// to the reference — normalized and correlation-scored — so absolute pitch and
// octave (male reciter vs any user) don't matter.

/** Map a pitch in Hz to a 0..1 ribbon height over a plausible vocal range. */
export function pitchToLevel(hz: number): number {
  const lo = Math.log2(90); // ~F#2
  const hi = Math.log2(520); // ~C5
  const v = (Math.log2(hz) - lo) / (hi - lo);
  return Math.max(0, Math.min(1, v));
}

/** Start mic capture; calls [onPitch] ~20×/s with a detected Hz (or null when
 *  unvoiced). Returns a stop() that releases the mic. Throws if permission is
 *  denied / no mic. */
export async function startPitchCapture(onPitch: (hz: number | null) => void): Promise<() => void> {
  const stream = await navigator.mediaDevices.getUserMedia({ audio: true });
  const Ctx = window.AudioContext || (window as unknown as { webkitAudioContext: typeof AudioContext }).webkitAudioContext;
  const ctx = new Ctx();
  const src = ctx.createMediaStreamSource(stream);
  const analyser = ctx.createAnalyser();
  analyser.fftSize = 2048;
  src.connect(analyser);
  const buf = new Float32Array(analyser.fftSize);
  let raf = 0;
  let last = 0;
  const tick = (t: number) => {
    if (t - last >= 50) { // ~20 Hz
      last = t;
      analyser.getFloatTimeDomainData(buf);
      onPitch(autoCorrelate(buf, ctx.sampleRate));
    }
    raf = requestAnimationFrame(tick);
  };
  raf = requestAnimationFrame(tick);
  return () => {
    cancelAnimationFrame(raf);
    src.disconnect();
    stream.getTracks().forEach((t) => t.stop());
    ctx.close();
  };
}

/** Autocorrelation pitch detector (after cwilso/PitchDetect). Returns Hz or
 *  null when the frame is too quiet / unvoiced. */
export function autoCorrelate(buf: Float32Array, sampleRate: number): number | null {
  const SIZE = buf.length;
  let rms = 0;
  for (let i = 0; i < SIZE; i++) rms += buf[i] * buf[i];
  rms = Math.sqrt(rms / SIZE);
  if (rms < 0.01) return null; // too quiet

  let r1 = 0, r2 = SIZE - 1;
  const thres = 0.2;
  for (let i = 0; i < SIZE / 2; i++) if (Math.abs(buf[i]) < thres) { r1 = i; break; }
  for (let i = 1; i < SIZE / 2; i++) if (Math.abs(buf[SIZE - i]) < thres) { r2 = SIZE - i; break; }
  const b = buf.slice(r1, r2);
  const n = b.length;

  const c = new Array(n).fill(0);
  for (let i = 0; i < n; i++) for (let j = 0; j < n - i; j++) c[i] += b[j] * b[j + i];

  let d = 0;
  while (d < n - 1 && c[d] > c[d + 1]) d++;
  let maxval = -1, maxpos = -1;
  for (let i = d; i < n; i++) if (c[i] > maxval) { maxval = c[i]; maxpos = i; }
  let T0 = maxpos;
  if (T0 <= 0) return null;

  // Parabolic interpolation for a finer estimate.
  const x1 = c[T0 - 1] ?? 0, x2 = c[T0], x3 = c[T0 + 1] ?? 0;
  const a = (x1 + x3 - 2 * x2) / 2;
  const bb = (x3 - x1) / 2;
  if (a) T0 = T0 - bb / (2 * a);

  const hz = sampleRate / T0;
  if (hz < 70 || hz > 700) return null; // outside a plausible voice range
  return hz;
}

/** Resample an array to [n] points by linear interpolation. */
function resample(src: number[], n: number): number[] {
  if (src.length === 0) return new Array(n).fill(0);
  if (src.length === 1) return new Array(n).fill(src[0]);
  const out: number[] = [];
  for (let i = 0; i < n; i++) {
    const t = (i / (n - 1)) * (src.length - 1);
    const lo = Math.floor(t), hi = Math.min(src.length - 1, lo + 1);
    out.push(src[lo] + (src[hi] - src[lo]) * (t - lo));
  }
  return out;
}

function minMaxNorm(a: number[]): number[] {
  const lo = Math.min(...a), hi = Math.max(...a);
  if (hi - lo < 1e-6) return a.map(() => 0.5);
  return a.map((v) => (v - lo) / (hi - lo));
}

function pearson(a: number[], b: number[]): number {
  const n = Math.min(a.length, b.length);
  let ma = 0, mb = 0;
  for (let i = 0; i < n; i++) { ma += a[i]; mb += b[i]; }
  ma /= n; mb /= n;
  let num = 0, da = 0, db = 0;
  for (let i = 0; i < n; i++) {
    const x = a[i] - ma, y = b[i] - mb;
    num += x * y; da += x * x; db += y * y;
  }
  if (da < 1e-9 || db < 1e-9) return 0;
  return num / Math.sqrt(da * db);
}

/**
 * Score how closely the learner's melodic SHAPE followed the reference.
 * Both are normalized (min-max) and progress-aligned (resampled), then
 * correlated — so octave / absolute pitch don't matter, only the up-and-down
 * journey. Returns 0..100.
 */
export function shapeMatch(userPitchesHz: number[], referenceLevels: number[]): number {
  if (userPitchesHz.length < 6) return 0;
  const N = 64;
  const user = minMaxNorm(resample(userPitchesHz.map((h) => Math.log2(h)), N));
  const ref = resample(referenceLevels, N); // already 0..1
  const r = pearson(user, ref);
  return Math.round(Math.max(0, r) * 100);
}
